### DISCUSSION TITLE
Why this project doiesn't work?

### DISCUSSION THREAD

**Discussion by vega0** [antlr/antlr4#4262] [Q&A]
Created multiple .g4 grammar rules, but when i some combine it in InstrykciyaImportirovat i''ve got some error. 

https://i.imgur.com/ye5AMoV.png

separately works normally https://i.imgur.com/pFepiRM.png

but when i include it it throws errors. How to fix it? https://vk.com/s/v1/doc/q1_FgtlMwZvys8ZIOP6pgCnuHiTk1M74ebK9Q_p9X86hxRtosGI

https://i.imgur.com/Xb8D73M.png

See the problem with token analyzer, but why its corrupting?

**Comment by kaby76** [comment]
I believe it doesn't work is because you define a grammar Virajeniya that is combined grammar (it contains both lexer and parser rules), and VirajeniyaLexer that is a lexer grammar (only lexer rules). When you go to generate the parsers/lexers, there will be a collision of the .tokens files between Virajeniya.g4 and VirajeniyaLexer.g4. You really should never define a "combined grammar"/"lexer grammar" pair in which there's going to be a generated file collision. In the following Bash shell, notice the sizes change for file VirajeniyaLexer.tokens depending on the order processed by the Antlr tool.
```
$ antlr4 Virajeniya.g4
warning(184): Virajeniya.g4:17:0: One of the token SLASH_PUTI values unreachable. / is always overlapped by token Deleniye
warning(125): Virajeniya.g4:7:24: implicit definition of token ZNAKI_A_Z_ in parser
warning(125): Virajeniya.g4:7:37: implicit definition of token ZNAKI_A_YA in parser
05/09-12:46:25 /c/Users/Kenne/Downloads/zz
$ ls -l Virajeniya
Virajeniya.g4                VirajeniyaBaseListener.java  VirajeniyaLexer.java         VirajeniyaParser.java
Virajeniya.interp            VirajeniyaLexer.g4           VirajeniyaLexer.tokens
Virajeniya.tokens            VirajeniyaLexer.interp       VirajeniyaListener.java
05/09-12:46:25 /c/Users/Kenne/Downloads/zz
$ ls -l Virajeniya*.tokens
-rw-r--r-- 1 Kenne Kenne 1017 May  9 12:46 Virajeniya.tokens
-rw-r--r-- 1 Kenne Kenne  987 May  9 12:46 VirajeniyaLexer.tokens
05/09-12:47:05 /c/Users/Kenne/Downloads/zz
$ antlr4 VirajeniyaLexer.g4
05/09-12:47:31 /c/Users/Kenne/Downloads/zz
$ ls -l Virajeniya*.tokens
-rw-r--r-- 1 Kenne Kenne 1017 May  9 12:46 Virajeniya.tokens
-rw-r--r-- 1 Kenne Kenne  137 May  9 12:47 VirajeniyaLexer.tokens
05/09-12:47:44 /c/Users/Kenne/Downloads/zz
```
In addition, you don't import nor use tokenVocab for the lexer grammar. This grammar cannot work. You should really use the Antlr tool, and see what the error messages the tool produces.

I recommend that you merge the lexer rules in VirajeniyaLexer.g4 into Virajeniya.g4. You need to read up on how to make a split grammar.

In addition, these grammars are overly complicated. You have four "top-level" grammars. I would highly recommend that you have one grammar with multiple entry points, and use a naming convention if necessary to avoid symbol collisions between the various grammars.
![graphviz (4)](https://github.com/antlr/antlr4/assets/11389451/0a6af1e2-2d63-42e5-b871-4e274dbe9c40)

On the use of `EOF`.
```
chislo: znakChisla? (desyaterichnoyeChislo | (binarnoyeChislo
	| tritichnoye
	| chetverichnoye
	| pyaterichnoye
	| shesterichnoyeChislo
	| semirichnoye
	| vosmerichnnoye
	| devyaterichnoyeChislo
	| shestnadcatirichnoye
)) EOF?;
```
I have never seen anything like this: `EOF?`. A start rule should be `EOF` terminated, and not optional. It forces the parser to consume all input.

**Reply by vega0** [reply]
How to avoid all collisions?

**Reply by vega0** [reply]
What does it mean? Token range : 

```
InstrykciyaVizovObiekta.g4 -o F:\eclipse-workspace\Российский ЯП\target\generated-sources\antlr4 -listener -no-visitor -lib F:\eclipse-workspace\Российский ЯП -encoding UTF-8
Exception in thread "main" java.util.NoSuchElementException: token index 56 out of range 0..52
	at org.antlr.runtime.BufferedTokenStream.get(BufferedTokenStream.java:154)
	at org.antlr.v4.parse.GrammarToken.getLine(GrammarToken.java:57)
	at org.antlr.runtime.tree.CommonTreeNodeStream.hasPositionInformation(CommonTreeNodeStream.java:208)
	at org.antlr.runtime.tree.CommonTreeNodeStream.remove(CommonTreeNodeStream.java:116)
	at org.antlr.runtime.misc.LookaheadStream.consume(LookaheadStream.java:97)
	at org.antlr.runtime.tree.TreeParser.matchAny(TreeParser.java:118)
	at org.antlr.v4.parse.BlockSetTransformer.topdown(BlockSetTransformer.java:328)
	at org.antlr.v4.parse.BlockSetTransformer.topdown(BlockSetTransformer.java:24)
	at org.antlr.runtime.tree.TreeRewriter$2.rule(TreeRewriter.java:111)
	at org.antlr.runtime.tree.TreeRewriter.applyOnce(TreeRewriter.java:61)
	at org.antlr.runtime.tree.TreeRewriter$1.pre(TreeRewriter.java:93)
	at org.antlr.runtime.tree.TreeVisitor.visit(TreeVisitor.java:56)
	at org.antlr.runtime.tree.TreeRewriter.downup(TreeRewriter.java:97)
	at org.antlr.runtime.tree.TreeRewriter.downup(TreeRewriter.java:86)
	at org.antlr.v4.tool.GrammarTransformPipeline.reduceBlocksToSets(GrammarTransformPipeline.java:89)
	at org.antlr.v4.analysis.LeftRecursiveRuleTransformer.translateLeftRecursiveRule(LeftRecursiveRuleTransformer.java:149)
	at org.antlr.v4.analysis.LeftRecursiveRuleTransformer.translateLeftRecursiveRules(LeftRecursiveRuleTransformer.java:96)
	at org.antlr.v4.semantics.SemanticPipeline.process(SemanticPipeline.java:94)
	at org.antlr.v4.Tool.processNonCombinedGrammar(Tool.java:400)
	at org.antlr.v4.Tool.process(Tool.java:387)
	at org.antlr.v4.Tool.processGrammarsOnCommandLine(Tool.java:346)
	at org.antlr.v4.Tool.main(Tool.java:193)

BUILD SUCCESSFUL
Total time: 1 second(s)
```

**Reply by vega0** [reply]
merged back lexers and parsers into one files.

**Reply by vega0** [reply]
https://vk.com/doc588434329_659512912

**Reply by kaby76** [reply]
> What does it mean? Token range :
> 
> ```
> Exception in thread "main" java.util.NoSuchElementException: token index 56 out of range 0..52
> ...
> 	at org.antlr.v4.Tool.main(Tool.java:193)
> 
> BUILD SUCCESSFUL
> Total time: 1 second(s)
> ```
After lexing, you have a token stream, and it is only 53 tokens in size. `org.antlr.runtime.BufferedTokenStream.get(int i)` accesses this token stream with a zero-integer base. You are trying to access past the length of the token stream, e.g., i = 53.

**Reply by vega0** [reply]
is it possible to increase size of token pool?

**Reply by kaby76** [reply]
I assume you are editing the token stream. A [TokenStream](https://www.antlr.org/api/Java/org/antlr/v4/runtime/TokenStream.html) interface has read-only methods. You will need to override [BufferedTokenStream](https://www.antlr.org/api/Java/org/antlr/v4/runtime/BufferedTokenStream.html) for your tree rewriter. Or, use [TokenStreamRewriter](https://www.antlr.org/api/Java/org/antlr/v4/runtime/TokenStreamRewriter.html).

But, I wouldn't even use an Antlr parse tree for tree rewriting because a token has an [index](https://github.com/antlr/antlr4/blob/8dcc6526cfb154d688497f31cf1e0904801c6df2/runtime/Java/src/org/antlr/v4/runtime/CommonToken.java#L68) of where it is in the token stream. Anytime you change the token stream, you will need to update that for all token in the token stream. There are other memoized values in the rest of the parse tree data structure. It is not a good data structure for tree rewriting.

**Reply by vega0** [reply]
Why tipBulevo: input 'boolean' doesn't match? 

https://pastebin.com/g6TDXyGv
